internal/k8s/controller.go

```diff
@@ -153,9 +154,105 @@ func isRouteUpdated(oldRoute, newRoute *routev1.Route) bool {
 	return false
 }
 
+func isNamespaceUpdated(oldNS, newNS *corev1.Namespace) bool {
+	if oldNS.ResourceVersion == newNS.ResourceVersion {
+		return false
+	}
+	oldLabelHash := utils.Hash(utils.Stringify(oldNS.Labels))
+	newLabelHash := utils.Hash(utils.Stringify(newNS.Labels))
+	if oldLabelHash != newLabelHash {
+		return true
+	}
+	return false
+}
+func AddIngressFromNSToIngestionQueue(numWorkers uint32, c *AviController, namespace string, msg string) {
+	ingObjs, err := utils.GetInformers().IngressInformer.Lister().ByNamespace(namespace).List(labels.Set(nil).AsSelector())
+	if err != nil {
+		utils.AviLog.Errorf("NS to ingress queue add: Error occured while retrieving ingresss for namespace: %s", namespace)
+		return
+	}
+	for _, ingObj := range ingObjs {
+		key := utils.Ingress + "/" + utils.ObjKey(ingObj)
+		bkt := utils.Bkt(namespace, numWorkers)
+		c.workqueue[bkt].AddRateLimited(key)
+		utils.AviLog.Debugf("key: %s, msg: %s for namespace: %s", key, msg, namespace)
+	}
+
+}
+
+/*
+ * Namespace Add event: will be called during each boot or newNS added. In add event
+ * handler, just add valid namespaces as Ingress handling, present in namespace, will be done
+ * during fullk8sync for reboot case or individual ingress handler called for ingress add event
+ * (For second case, user will add namespace first and then ingress, so just validating namespace
+ * should be enough)
+ */
+/* Namespace Delete event: no op. Let individual event handler take care
+ */
+/* Namespace update event:  2 cases to handle : NS label changed from 1) valid to invalid --> Call ingress delete
+ * 2) invalid to valid --> Call ingress add
+ */
+
+func AddNameSpaceEventHandler(numWorkers uint32, c *AviController) cache.ResourceEventHandler {
+	namespaceEventHandler := cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			if c.DisableSync {
+				return
+			}
+			ns := obj.(*corev1.Namespace)
+			nsLabels := ns.GetLabels()
+			nsFilterObj := utils.GetGlobalK8NSObj()
+			if utils.NSFilterFunction(ns.GetName(), nsFilterObj, nsLabels, false) {
+				utils.AddNamespace(ns.GetName(), nsFilterObj)
+				utils.AviLog.Debugf("NS Add event: Namespace passed filter: %s", ns.GetName())
+			} else {
+				//Case: previoulsy deleted valid NS, added back with no labels or invalid labels but nsList contain that ns
+				utils.AviLog.Debugf("NS Add event: Namespace did not pass filter: %s", ns.GetName())
+				if utils.NSFilterFunction(ns.GetName(), nsFilterObj, nil, true) {
+					utils.AviLog.Debugf("Ns Add event: Deleting previous valid namespace: %s from valid NS List", ns.GetName())
+					utils.DeleteNamespace(ns.GetName(), nsFilterObj)
```

Comment:
If this deletes the namespace from the filter, then let's rename it to `DeleteNamespaceFromFilter`. What do you think?